core/modules/replica/MessengerConnector.cc

```diff
@@ -0,0 +1,640 @@
+                            // a problem with the request, remove the request from all data structures and
+                            // restart the communication.
+
+                            request2notify = _currentRequest;
+
+                            _id2request.erase(_currentRequest->id);
+                            _currentRequest = nullptr;
+
+                            restart(lock);
+
+                        } else {
+
+                            // Finally, success!
+
+                            // Notify a caller about the successfull completion of the operation,
+                            // remove this request from all data structures and initiate a new request.
+
+                            request2notify = _currentRequest;
+                            request2notify->success = true;
+
+                            _id2request.erase(_currentRequest->id);
+                            _currentRequest = nullptr;
+
+                            sendRequest(lock);
+                        }
+                    }
+                }
+            }
+
+        } else {
+
+            // The request submission had a chance to finish (successfully or not)
+            // before the cancellation attempt was made. So, we just forget about
+            // this request and send no notification to a caller.
+            //
+            if (ec) restart(lock);
+            else    sendRequest(lock);
+        }
+    }
+
+    // Sending notifications (if requsted) outsize the lock guard
+
+    if (request2notify) request2notify->parseAndNotify();
+}
+
+boost::system::error_code MessengerConnector::syncReadFrame(util::Lock const& lock,
+                                                            replica::ProtocolBuffer& buf,
+                                                            size_t& bytes) {
+    size_t const frameLength = sizeof(uint32_t);
+    buf.resize(frameLength);
+
+    boost::system::error_code ec;
+    boost::asio::read(
+        _socket,
+        boost::asio::buffer(
+            buf.data(),
+            frameLength
+        ),
+        boost::asio::transfer_at_least(frameLength),
+        ec
+    );
+    LOGS(_log, LOG_LVL_DEBUG, context() << "syncReadFrame"
+         << "  _currentRequest->id=" << (_currentRequest ? _currentRequest->id : "")
+         << " error_code=" << ec);
+
+    if (not ec) bytes = buf.parseLength();
+    return ec;
+}
+
+boost::system::error_code MessengerConnector::syncReadVerifyHeader(util::Lock const& lock,
+                                                                   replica::ProtocolBuffer& buf,
+                                                                   size_t bytes,
+                                                                   std::string const& id) {
+
+    boost::system::error_code const ec =
+        syncReadMessageImpl(lock,
+                            buf,
+                            bytes);
```

Comment:
so many lines. Is your editor window very narrow?

Reply:
It's more convenient to read the code written in this way. I follow this pattern in header files (declarations) and the implementation files.